### DISCUSSION TITLE
MCP23S08 inputs with counter support

### DISCUSSION THREAD

**Discussion by Frank-1964** [arendst/Tasmota#20988] [Support request]
Hi everybody,

I own a KMP PRODUINo ESP32 board. This board has a MCP23S08 with 4 gpios as relay outputs and 4 gpios as inputs.

After I change the file tasmota/tasmota_xdrv_driver/xdrv_67_mcp23xxx.ino (see https://github.com/arendst/Tasmota/pull/20971) the relay are working properly.

Now I want to get the inputs working with sml (counter) support. At this time I have no idea how to get the counter stuff working. Can somebody push me in the right direction?

Thanks,
Frank

**Comment by sfromis** [comment]
Configuring Counter inputs is not a feature of the driver for the multiplexer, which makes sense as the Counter support is based on quick reaction to interrupts, instead of I2C and polling. If you have a "real" gpio not hobbled by the multiplexer chip, that would be a simple approach.

**Reply by Frank-1964** [reply]
Hmmm ... would you say that it is not possible to use counter support to the MCP23S08 inputs with the current framefork?

**Reply by sfromis** [reply]
I'd say that current Counter support, aiming to support relatively high frequencies, is not viable for port extender inputs. And using "Counter" instead of "counter" was to hint at this **not** being about the Smart Meter Interface, but about the basic Tasmota feature with straight pulse counts.

For much lower frequencies, it could be feasible to config the port extender input as a Switch, and then use rules to issue the Counter command to increment the pulse count.

What workarounds may (or not) be feasible for the Smart Meter Interface - I have no experience with the details of this feature, and can't really answer here. But **if** the Counter command works for those, same method might apply. I'm not really expecting this to be the case; just don't know much about how it works there.

Seeing a page like https://kmpelectronics.eu/products/prodino-esp32-v1/ suggests that you have better options than that port extender, as the page talks about a Grove connector with gpios, and additional internal connectors with even more gpio pins. Obviously, make sure that things are electrically compatible.

**Comment by stsch83** [comment]
Hi Frank,

can you explain how to get the kmp io extender work.

BR,
Stefan

**Reply by Frank-1964** [reply]
The IO expander MCP23S08 is part of the current tasmota firmware (14.0.0). I use this template:

{"NAME":"KMP PRODINo ESP32","GPIO":[1376,0,0,0,0,0,0,0,0,0,0,0,0,0,736,672,0,0,1312,704,0,0,0,0,0,0,0,0,10272,3840,0,0,10240,0,0,0],"FLAG":0,"BASE":1}

and use this configuration:

GPIO18	SPI CLK
GPIO19	SPI MISO
GPIO23	SPI MOSI
GPIO32	MCP23Sxx CS
GPIO33	Output Hi
GPIO36	MCP23xxx Int

Frank

**Reply by stsch83** [reply]
this will not work for me. Is there something more to do?

**Reply by Frank-1964** [reply]
Not that I know of.
What kind of hardware do you own?

**Reply by stsch83** [reply]
i have KMP PRODINo ESP32 Ethernet

**Reply by stsch83** [reply]
Ehternet (EthType 8) and the LED is working.

**Reply by Frank-1964** [reply]
Hmmm .... I don't configured more than said. 
Yes, the LED is working too. I don't use the eth port.

**Reply by Frank-1964** [reply]
Can you please send the boot log from the console?

**Reply by stsch83** [reply]
00:00:00.002 HDW: ESP32-D0WD-V3 v3.1 
00:00:00.044 UFS: FlashFS mounted with 304 kB free
00:00:00.068 CFG: Loaded from File, Anzahl 12
00:00:00.079 QPC: Count 1
00:00:00.083 SPI: Hardware using GPIO18(CLK), GPIO23(MOSI) and GPIO19(MISO)
00:00:00.140 BRY: Berry initialized, RAM used 3884 bytes
00:00:00.156 Projekt tasmota - Tasmota Version 14.0.0(release-tasmota32)-3_0_0(2024.05.14 13:58:38)
00:00:01.000 WIF: verbinden mit AP1 wirschlegels IoT Channel 1 BSSId 7A:5A:B0:45:A2:EE in Modus HT20 als tasmota-242A30-2608...
00:00:03.375 WIF: verbunden
00:00:03.631 HTP: Webserver aktiv bei tasmota-242A30-2608 mit IP-Adresse 192.168.100.229
19:41:36.460 RSL: INFO1 = {"Info1":{"Module":"KMP PRODINo ESP32","Version":"14.0.0(release-tasmota32)","FallbackTopic":"cmnd/DVES_242A30_fb/","GroupTopic":"cmnd/tasmotas/"}}
19:41:36.473 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-242A30-2608","IPAddress":"192.168.100.229","IP6Global":"","IP6Local":"fe80::ad1:f9ff:fe24:2a30%st1"}}
19:41:36.485 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":4}}
19:41:36.497 RSL: RESULT = {"POWER":"OFF"}
19:41:36.498 RSL: POWER = OFF
19:41:37.936 QPC: Reset
19:41:39.921 RSL: STATE = {"Time":"2024-05-26T19:41:39","Uptime":"0T00:00:08","UptimeSec":8,"Heap":150,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":26,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"POWER":"OFF","Dimmer":10,"Color":"1A1A1A","HSBColor":"0,0,10","Channel":[10,10,10],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"wirschlegels IoT","BSSId":"7A:5A:B0:45:A2:EE","Channel":1,"Mode":"HT20","RSSI":20,"Signal":-90,"LinkCount":1,"Downtime":"0T00:00:04"}}

**Reply by Frank-1964** [reply]
00:00:00.002 HDW: ESP32-D0WD v1.0 
00:00:00.033 UFS: FlashFS mounted with 304 kB free
00:00:00.052 CFG: Loaded from File, Count 574
00:00:00.063 QPC: Count 1
00:00:00.068 SPI: Hardware using GPIO18(CLK), GPIO23(MOSI) and GPIO19(MISO)
00:00:00.098 SPI: MCP23S08 found at CS1
00:00:00.171 BRY: Berry initialized, RAM used 3884 bytes
00:00:00.189 Project tasmota - Tasmota Version 14.0.0(release-tasmota32)-3_0_0(2024-05-14T13:54:09)

That is mine ... the line with MCSP23S08 found is missing in your log. Maybe there are hardware differences? My log shows a ESP32-D0WD v1.0, youors is a V3 ...

**Reply by Frank-1964** [reply]
Wait ..... maybe you are missing the mcp23x.dat file?

**Reply by stsch83** [reply]
i found this Schematic. https://kmpelectronics.eu/wp-content/uploads/2019/12/ProDINoESP32-ETHv02-Schematic.pdf

**Reply by stsch83** [reply]
i dont have this file

**Reply by sfromis** [reply]
Did you do this? https://tasmota.github.io/docs/MCP230xx/#mode-2

**Reply by stsch83** [reply]
i tried before. but i was also not working

**Reply by Frank-1964** [reply]
You need a file mcp23x.dat with this content:

{"NAME":"KMP PRODINo ESP32 MCP23S08","GPIO":[160,161,162,163,224,225,226,227]}

and put it to the esp23 with the 'manage filesystem' tool

**Reply by stsch83** [reply]
Now it looks like this:

00:00:00.002 HDW: ESP32-D0WD-V3 v3.1 
00:00:00.028 UFS: FlashFS mounted with 304 kB free
00:00:00.043 CFG: Loaded from File, Anzahl 45
00:00:00.054 QPC: Count 1
00:00:00.058 SPI: Hardware using GPIO18(CLK), GPIO23(MOSI) and GPIO19(MISO)
00:00:00.082 SPI: MCP23S08 found at CS1
00:00:00.150 BRY: Berry initialized, RAM used 3884 bytes

**Reply by stsch83** [reply]
but i can not see any input changes and i can not switch the relays

**Reply by stsch83** [reply]
what does 160,161,162,163 (0xA0 ...) and 224,225,226,227 (0xE0 ...) in the mcp23x.dat means?

**Reply by Frank-1964** [reply]
https://tasmota.github.io/docs/MCP230xx/#mode-2 describes that

**Reply by stsch83** [reply]
are this the addresses of the inputs and outputs of the mcp23s08?

**Reply by stsch83** [reply]
i had a little mistake in my mcp23x.dat file. Now it's working. Thx a lot.

**Reply by stsch83** [reply]
Is it normal, when the input is On the relay is also On?

**Reply by Frank-1964** [reply]
Congrats :-)
Yes, its the same here. I don't use the input currently, so I did not take notice of that until now :-)

**Reply by Frank-1964** [reply]
You can detatch the relays from the input with setOption ...
